src/base/baseBrushTool.js

```diff
@@ -0,0 +1,367 @@
+import external from './../externalModules.js';
+import EVENTS from './../events.js';
+import baseTool from './../base/baseTool.js';
+// Utils
+import KeyboardController from '../fancy-tools/shared/KeyboardController.js';
+import isToolActive from '../fancy-tools/shared/isToolActive.js';
+import { getNewContext } from '../util/drawing.js';
+import { COLOR_MAP_ID } from '../stateManagement/brushToolColors.js';
+
+export default class extends baseTool {
+  constructor ({
+    name,
+    strategies,
+    defaultStrategy,
+    configuration,
+    supportedInteractionTypes
+  }) {
+    super({
+      name,
+      strategies,
+      defaultStrategy,
+      configuration,
+      supportedInteractionTypes
+    });
+
+    this.isBrushTool = true;
+    this.hasCursor = true;
+    this._referencedToolData = 'brush';
+
+    this._drawingMouseUpCallback = this._drawingMouseUpCallback.bind(this);
+  }
+
+  //===================================================================
+  // Abstract Methods - Must be implemented.
+  //===================================================================
+
+  /**
+  * Helper function for rendering the brush.
+  *
+  * @abstract
+  * @param {Object} evt - The event.
+  */
+  renderBrush (evt) {
```

Comment:
I know very little about the brush tool. Is there a reason to provide this method AND the `renderToolData` method? Is it a way to separate brush rendering logic from annotations the tool may also create (IE. textboxes holding calculations related to what's been painted)?

Reply:
I agree should be refactored. My original rationale for doing it this way: trying to fit a rather new methodology into the current framework.

I only wanted one brush tool to do all the rendering of the actual mask. There is one common data file per slice that "all" brushes manipulate (currently only 2 tools, but we envisage at least 2 more). Currently the brushTool overrides renderToolData to draw the actual mask, then calls renderBrush. The eraser uses the default renderToolData which just wraps renderBrush.

The downside of this approach is that currently the brushTool needs to be enabled for the brushEraserTool to function, which isn't too nonsensical for now. However, once we have an adaptive brush, perhaps a user would like to enable e.g. only adaptiveBrushTool and brushEraserTool.

I think the solution for this is as follows:
- renderToolData just renders the cursor (leave the wrapper for easy to follow naming? idk).
- in eventDispatchers/onImageRendered:
  - filters tools by (instanceof baseBrushTool).
  - if # baseBrushTools > 0, render brush data.

I'm happy to work on this refactor, but will wait for an opinion on this idea first.

Reply:
I "think" I follow, but @swederik may have better thoughts about how to proceed. Times like this I wish we shared an office so we could whiteboard.

Reply:
I can explain it on Thursday more clearly :).